src/kixi/datastore/segmentation/inmemory.clj

```diff
@@ -0,0 +1,170 @@
+            [clojure-csv.core :as csv :refer [parse-csv write-csv]]
+            [kixi.datastore.segmentation :as seg :refer [Segmentation]]
+            [kixi.datastore.communications :refer [Communications] :as comms]
+            [kixi.datastore.filestore :as kdfs]
+            [kixi.datastore.metadatastore :as ms]
+            [taoensso.timbre :as timbre :refer [error info infof]]
+            [kixi.datastore.communications :as c]))
+
+(defn uuid
+  []
+  (str (java.util.UUID/randomUUID)))
+
+(defn retrieve-file-to-local
+  [filestore id]
+  (let [^java.io.File file (java.io.File/createTempFile id ".tmp")]
+    (if-let [src (kdfs/retrieve filestore id)]
+      (do (bs/transfer src
+                       file)
+          file)
+      {::seg/reason :file-not-found
+       ::seg/cause id})))
+
+(defn new-segment-data
+  [headers value]
+  (let [id (uuid)
+        ^java.io.File file (java.io.File/createTempFile id ".tmp")
+        os (io/output-stream file)
+        header-line (write-csv [headers])
+        _ (bs/transfer header-line
+                       os
+                       {:close? false})]
+    {:value value
+     :file file
+     :id id
+     :lines 1
+     :size-bytes (alength (.getBytes ^String header-line))
+     :output-stream os}))
+
+(defn index-of
+  [coll value]
+  (first
+   (keep-indexed
+    #(when (= value %2)
+       %1)
+    coll)))
+
+(defn segmentate-file-by-column-values
+  [column-name file]
+  (with-open [contents (io/reader file)]
+    (let [parser (parse-csv contents :strict true)
+          header-line (first parser)
+          dex (index-of header-line column-name)]
+      (if dex
+        (->> (rest parser)
+             (reduce 
+              (fn [value->segment-data line]
+                (let [value (nth line dex)
+                      segment-data (get value->segment-data value
+                                        (new-segment-data header-line value))
+                      line-csv (write-csv [line])]
+                  (bs/transfer line-csv
+                               (:output-stream segment-data)
+                               {:close? false
+                                :append? true})                  
+                  (assoc value->segment-data
+                         value
+                         (-> segment-data
+                             (update :lines
+                                     inc)
+                             (update :size-bytes
+                                     (partial + (alength (.getBytes ^String line-csv))))))))
+              {})
+             vals
+             (map (fn [segment-data]
+                    (.close ^java.io.OutputStream (:output-stream segment-data))
+                    segment-data)))
+        {::seg/reason :invalid-column
+         ::seg/cause column-name}))))
+
+(defn upload-segment
+  [filestore communications]
+  (fn [basemetadata request segment-data]
+    (let [_ (bs/transfer (:file segment-data)
```

Comment:
nope, i just kept this one to make it clear that there's a lot of fire and forget. It's time has probably come though